shachain/element.go

```diff
@@ -0,0 +1,134 @@
+package shachain
+
+import (
+	"crypto/sha256"
+	"errors"
+
+	"github.com/roasbeef/btcd/wire"
+)
+
+// Element represent the entityt which contains the hash and
+// correponding to it index. By comparing two indexes we may change hash in
+// such way to derive another element.
+type Element struct {
+	index index
+	hash  *wire.ShaHash
+}
+
+// NewElementFromStr creates new element by the given hash string.
+func NewElementFromStr(s string, index index) (*Element, error) {
+	hash, err := hashFromString(s)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Element{
+		index: index,
+		hash:  hash,
+	}, nil
+}
+
+// derive is used to get the one shachain element from another by applying
+// bits flipping and hash operation based on index.
+func (e *Element) derive(to index) (*Element, error) {
+	positions, err := e.index.derive(to)
+	if err != nil {
+		return nil, err
+	}
+
+	buf := e.hash.Bytes()
+	for _, position := range positions {
+		// Flip the bit and then hash the current state.
+		h := sha256.Sum256(changeBit(buf, position))
+		buf = h[:]
+	}
+
+	hash, err := wire.NewShaHash(buf)
+	if err != nil {
+		return nil, err
+	}
+
+	return &Element{
+		index: to,
+		hash:  hash,
+	}, nil
+}
+
+// isEqual checks elements equality.
+func (first *Element) isEqual(second *Element) bool {
+	return (first.index == second.index) && first.hash.IsEqual(second.hash)
+}
+
+const (
+	// maxHeight is used to determine the the maximum allowable index and
+	// the lenght of array which should be stored in order to have
+	// ability to derive all previous hashes by index (this array also knows
+	// as buckets)
+	maxHeight uint8 = 48
+
+	// rootIndex is an index which corresponds to the root hash.
+	rootIndex index = 0
+)
+
+// startIndex is an index of first element.
+var startIndex index = (1 << maxHeight) - 1
+
+// index is an number which identifies the hash number and serve as the way to
+// determine which operation under hash we should made in order to derive one
+// hash from another. index initialized with start index value and than
+// decreases down to zero.
+type index uint64
+
+// NewIndex is used to create index instance. The inner operations with
+// index implies that index decresiang from some max number to zero, but for
+// simplicity and backward compatibility with previous logic it was transformed
+// to work in opposite way.
+func NewIndex(v uint64) index {
+	return startIndex - index(v)
+}
+
+// derive function checks that 'to' index is deriavable from 'from' index
+// by checking the indexes prefixes and than returns bit positions at which
+// zeros should to be changed to ones in order indexes becomes the same. This
+// bits set is needed in order to derive one hash from another.
+func (from index) derive(to index) ([]uint8, error) {
+	positions := make([]uint8, 0)
+	if from == to {
+		return positions, nil
+	}
+
+	//	+ --------------- +
```

Comment:
Hmm, I'm not sure if this diagram aides my understanding of the method used to derive the series of bit-flips (then hashes) required to derive the next element. Do you think you could add a comment above or below the diagram further explaining the algorithm? Thanks!